File: mko.Asp/HtmlCtrl/HtmlNumBox.js
/// <reference name="MicrosoftAjax.js" />
/*
    (c) MKIT Martin Korneffel IT-Beratung/Softwareentwicklung
    Stuttgart, den 3.4.2012

    Funktionen der Clientkomponente von NumBox

    Eingebettete Resource:
     => Eigenschaften/Buildvorgang := EmbeddedResource
     => AssemblyInfo/[assembly: System.Web.UI.WebResource("mkoIt.HtmlNumBox.js", "text/javascript")]
*/

//---------------------------------------------------------------------------------
// Namensraum Registrieren

Type.registerNamespace("mkoIt.Asp.Html.Ajax");

//---------------------------------------------------------------------------------
// Klasse NumBox anlegen
Sys.Debug.trace("Deklaration der NumBox- Clientklasse");

// Konstruktor
mkoIt.Asp.Html.Ajax.NumBox = function (element) {
    mkoIt.Asp.Html.Ajax.NumBox.initializeBase(this, [element]);

    this._numPadId = "";
    Sys.Debug.assert($get(numPadId) != null, "Die NumPad- Dom- Struktur für eine NumBox existiert nicht", true);

    this._d0 = "";
    this._d1 = "";
    this._d2 = "";
    this._d3 = "";
    this._d4 = "";
    this._d5 = "";
    this._d6 = "";
    this._d7 = "";
    this._d8 = "";
    this._d9 = "";
    this._hide = "";
    this._back = "";

}

// Prototype
mkoIt.Asp.Html.Ajax.NumBox.prototype = {

    initialize: function () {
        mkoIt.Asp.Html.Ajax.NumBox.callBaseMethod(this, "initialize");
    },

    get_NumPadId: function () {
        return this._numPadId;
    },

    set_NumPadId: function (value) {
        this._numPadId = value;
    },

    // NumPad Tasten definieren
    set_d0: function (value) {
        $addHandler($get(value), "click", this.AddD0);
    },

    set_d1: function (value) {
        $addHandler($get(value), "click", this.AddD1);
    },

    set_d2: function (value) {
        $addHandler($get(value), "click", this.AddD2);
    },


    ShowNumPad: function () {
        $get(this._numPadId).setAttribute("style", "visibility: visible");
    },

    HideNumPad: function () {
        $get(this._numPadId).setAttribute("style", "visibility: collapse");
    },

    AddDigit: function (digit) {
        if (this.get_element() && !this.get_element().disabled) {
            var input = this.get_element().getAttribute("value");
            input += digit;
            this.get_element().setAttribute("value", input);
        }
    },

    AddD0: function (e) {
        this.AddDigit("0");
    },




    BackSpace: function () {
        var input = this.get_element().getAttribute("value");
        if (input.length > 0) {
            input = input.substring(0, input.length - 1);
        }
        this.get_element().setAttribute("value", input);
    },

    Clear: function () {
        this.get_element().setAttribute("value", "");
    }
}

// Typ registrieren
mkoIt.Asp.Html.Ajax.NumBox.registerClass("mkoIt.Asp.Html.Ajax.NumBox", Sys.UI.Control);



//---------------------------------------------------------------------------------
// Klasse NumBoxUtils anlegen
Sys.Debug.trace("Deklaration der NumBoxUtils- Clientklasse");

// Konstruktor
mkoIt.Asp.Html.Ajax.NumBoxUtils = function () {
    mkoIt.Asp.Html.Ajax.NumBoxUtils.initializeBase(this);

    this._textBoxId = "";
    this._numPadId = "";

}

// Prototype
mkoIt.Asp.Html.Ajax.NumBoxUtils.prototype = {

    initialize: function () {
        mkoIt.Asp.Html.Ajax.NumBoxUtils.callBaseMethod(this, "initialize");
    },

    get_TextBoxId: function () {
        return this._textBoxId;
    },

    set_TextBoxId: function (value) {
        Sys.Debug.assert($get(value) !== null, "Die TextBox für eine NumBox existiert nicht", true);
        this._textBoxId = value;
    },

    get_NumPadId: function () {
        return this._numPadId;
    },

    set_NumPadId: function (value) {
        Sys.Debug.assert($get(value) !== null, "Die NumPad- Struktur für eine NumBox existiert nicht", true);
        this._numPadId = value;
    },  


    ShowNumPad: function () {
        $get(this._numPadId).setAttribute("style", "visibility: visible");
    },

    HideNumPad: function () {
        $get(this._numPadId).setAttribute("style", "visibility: collapse");
    },

    AddDigit: function (digit) {
        if ($get(this._textBoxId) && !$get(this._textBoxId).disabled) {
            var input = $get(this._textBoxId).getAttribute("value");
            input += digit;
            $get(this._textBoxId).setAttribute("value", input);
        }
    },

    BackSpace: function () {
        var input = $get(this._textBoxId).getAttribute("value");
        if (input.length > 0) {
            input = input.substring(0, input.length - 1);
        }
        $get(this._textBoxId).setAttribute("value", input);
    },

    Clear: function () {
        $get(this._textBoxId).setAttribute("value", "");
    }
}

// Typ registrieren
mkoIt.Asp.Html.Ajax.NumBoxUtils.registerClass("mkoIt.Asp.Html.Ajax.NumBoxUtils", Sys.Component);


// Rückmelden, falls Script bereit geladen wurde
if (typeof (Sys) !== 'undefined') Sys.Application.notifyScriptLoaded();

File: mko.Asp/HtmlCtrl/HtmlNumBox.cs
//<unit_header>
//----------------------------------------------------------------
//
// Martin Korneffel: IT Beratung/Softwareentwicklung
// Stuttgart, den 23.3.2012
//
//  Projekt.......: mkoItAsp
//  Name..........: HtmlNumBox.cs
//  Aufgabe/Fkt...: Html- Fromularelement für nummerische Eingaben.
//                  bietet eine Tastatur für die Eingabe an.
//
//
//
//
//<unit_environment>
//------------------------------------------------------------------
//  Zielmaschine..: PC 
//  Betriebssystem: Windows XP mit .NET 4.0
//  Werkzeuge.....: Visual Studio 2005
//  Autor.........: Martin Korneffel (mko)
//  Version 1.0...: 
//
// </unit_environment>
//
//<unit_history>
//------------------------------------------------------------------
//
//  Version.......: 1.1
//  Autor.........: Martin Korneffel (mko)
//  Datum.........: 
//  Änderungen....: 
//
//</unit_history>
//</unit_header>        

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using System.Web;
using System.Web.UI;
using MsWebCtrls = System.Web.UI.WebControls;
using MsHtmlCtrls = System.Web.UI.HtmlControls;

using mkx = mkoIt.Xhtml.Xhtml;

namespace mkoIt.Asp.HtmlCtrl
{
    public partial class NumBox : DIV, IScriptControl
    {
        TextBox tbx;
        DIV divNumPad;        

        // ID- Generatoren für die Bausteine der NumBox
        string IdTbx()
        {
            return ID + "_Tbx";            
        }

        string IdBtnNumPadOpen()
        {
            return ID + "_BtnNumPadOpen";
        }

        string IdNumPad()
        {
            return ID + "_NumPad";
        }

        string IdNumPadKey(int digit)
        {
            return ID + "_NumPad_" + digit;
        }

        string IdNumPadBackSpace()
        {
            return ID + "_NumPad_BackSpace";
        }

        string IdNumPadHide()
        {
            return ID + "_NumPad_Hide";
        }

        string IdNumPadClear()
        {
            return ID + "_NumPad_Clear";
        }

        string IdNumPadKomma()
        {
            return ID + "_NumPad_Komma";
        }

        string IdNumPadPoint()
        {
            return ID + "_NumPad_Point";
        }

        string IdClientComponent()
        {
            return ClientID + "_ClientComponent";        
        }

        // Schnittstelle zur Clientkomponente 
        ClientComponentScriptBld myClientComponent;


        // Bausteine der Numbox
        TableCell CreateDigitCell(int digit)
        {
            return new TableCell(
                      new Button(IdNumPadKey(digit))
                      {
                          Text = digit.ToString(),
                          SetClientClick = myClientComponent.AddDigit(digit)
                      });
        }

        TableCell CreateBackSpaceCell()
        {
            return new TableCell(
                new Button(IdNumPadBackSpace())
                {
                    Text = "D",
                    SetClientClick = myClientComponent.BackSpace()
                });
        }

        TableCell CreateClearCell()
        {
            return new TableCell(
                new Button(IdNumPadClear())
                {
                    Text = "CE",
                    SetClientClick = myClientComponent.Clear()
                });
        }

        TableCell CreateHideCell()
        {
            return new TableCell(
                new Button(IdNumPadHide())
                {
                    Text = "X",                    
                    SetClientClick = myClientComponent.HideNumPad(),

                    CssStyleBld = new Xhtml.Css.StyleBuilder() {
                        BackgroundColor = Xhtml.Css.Color.Red
                    }
                });
        }

        TableCell CreateKommaCell()
        {
            return new TableCell(
                new Button(IdNumPadKomma())
                {
                    Text = ",",
                    SetClientClick = myClientComponent.AddKomma()
                });
        }


        TableCell CreatePointCell()
        {
            return new TableCell(
                new Button(IdNumPadPoint())
                {
                    Text = ".",
                    SetClientClick = myClientComponent.AddPoint()
                });

        }

        TableRow CreateRow(int startDigit, TableCell ctrlKey)
        {
            return new TableRow(
                CreateDigitCell(startDigit),
                CreateDigitCell(startDigit + 1),
                CreateDigitCell(startDigit + 2),
                ctrlKey);
        }       

        // Konostruktor
        public NumBox(string ID)
        {
            CreateNumbox(ID);
        }

        public NumBox(string ID, out NumBox numBoxRef)
        {
            numBoxRef = this;
            CreateNumbox(ID);
        }

        private void CreateNumbox(string ID)
        {
            this.ID = ID;
            var ParentCssBld = CssStyleBld;
            myClientComponent = new ClientComponentScriptBld(IdClientComponent());

            Content = new Control[]{
                new TextBox(IdTbx(), out tbx)
                {
                    CssStyleBld = new Xhtml.Css.StyleBuilder() {
                        TextAlign = Xhtml.Css.TextAlign.Right
                    }
                },

                new Button(IdBtnNumPadOpen()){
                    Text = "#",
                    SetClientClick = myClientComponent.ShowNumPad()                    
                },

                new DIV(
                    IdNumPad(), 
                    out divNumPad,
                    new Control[] {
                        new Table(
                                    CreateRow(7, CreateHideCell()), 
                                    CreateRow(4, CreateBackSpaceCell()),
                                    CreateRow(1, CreateClearCell()),
                                    new TableRow(
                                            CreateDigitCell(0),
                                            CreatePointCell(),
                                            CreateKommaCell()
                                        )
                                )
                                {
                                    ID = this.ID + "TabNumPad",
                                    CssStyleBld = new Xhtml.Css.StyleBuilder()
                                    {
                                        TableBorderCollapse = Xhtml.Css.BorderCollapse.Collapse,
                                        Padding =  Xhtml.Css.Length.Pixel(3),
                                        Position = Xhtml.Css.Position.Fixed                                       
                                    }
                                }
                    })
                {
                    CssStyleBld = new Xhtml.Css.StyleBuilder() {
                        Visibility = Xhtml.Css.Visiblity.Collapse
                    }
                }
            };

        }

        /// <summary>
        /// Zugriff auf den Inhalt vom Server aus
        /// </summary>
        public string Text
        {
            get
            {
                return tbx.Text;
            }

            set
            {
                tbx.Text = value;
            }
        }
              

        public override void PreRenderReworking(object sender)
        {

            if (CssStyleBld != null && CssStyleBld.TextAlign == null)
                CssStyleBld.TextAlign = Xhtml.Css.TextAlign.Right;

            // Registrieren der Clientkomponente

            var sm = ScriptManager.GetCurrent(Page);

            if (sm == null)
                throw new HttpException("A ScriptManager control must exist on the current Page.");

            // Script einbinden
            sm.RegisterScriptControl(this);
        }

        protected override void Render(HtmlTextWriter writer)
        {
            base.Render(writer);

            // Anlegen der Clientinstanzen mit $create
            if (!this.DesignMode)
                ScriptManager.GetCurrent(Page).RegisterScriptDescriptors(this);
        }

        public EventHandler SetLoad
        {
            set
            {
                tbx.Load += new EventHandler(value);
            }
        }

        public EventHandler SetTextChanged
        {
            set
            {
                ((System.Web.UI.WebControls.TextBox)tbx).TextChanged += new EventHandler(value);
            }
        } 

        // Implementierung der IScriptControl- Schnittstelle

        IEnumerable<ScriptDescriptor> IScriptControl.GetScriptDescriptors()
        {
            // $create- Aufruf für die Instanzirieung der Componente beschreiben
            var descr = new ScriptComponentDescriptor("mkoIt.Asp.Html.Ajax.NumBoxUtils");

            descr.AddProperty("id", IdClientComponent());
            descr.AddProperty("TextBoxId", tbx.ClientID);
            descr.AddProperty("NumPadId", divNumPad.ClientID);            

            return new ScriptDescriptor[] { descr };
        }

        IEnumerable<ScriptReference> IScriptControl.GetScriptReferences()
        {
            // Referenz auf eingebettetes Script zurückgeben
            var reference = new ScriptReference()
            {
                Assembly = "mko.Asp",
                Name = "mko.Asp.HtmlNumBox.js"
            };

            return new ScriptReference[] { reference };
        }
    }
}

